pulsar-client-cpp/lib/Backoff.cc

```diff
@@ -17,24 +17,40 @@
  * under the License.
  */
 #include "Backoff.h"
-#include <algorithm>
 
 namespace pulsar {
 
-Backoff::Backoff(const TimeDuration& initial, const TimeDuration& max)
+Backoff::Backoff(const TimeDuration& initial, const TimeDuration& max, const TimeDuration& mandatoryStop)
         : initial_(initial),
           max_(max),
-          next_(initial) {
+          next_(initial),
+          mandatoryStopMade_(false),
+          mandatoryStop_(mandatoryStop),
+          seed_(time(NULL)) {
 }
 
 TimeDuration Backoff::next() {
     TimeDuration current = next_;
     next_ = std::min(next_ * 2, max_);
-    return current;
+
+    // Check for mandatory stop
+    if (initial_ == current) {
+        timeElapsedSinceDisconnection_ = boost::posix_time::milliseconds(0);
+    }
+    if (!mandatoryStopMade_ && timeElapsedSinceDisconnection_ + current > mandatoryStop_) {
+        current = std::max(initial_, mandatoryStop_ - timeElapsedSinceDisconnection_);
+        mandatoryStopMade_ = true;
+    }
+    timeElapsedSinceDisconnection_ += current;
```

Comment:
Given that the exact current time might be slightly different than the "sleep" time, I think we should be better use an absolute time instead of `timeElapsedSinceDisconnection_`. That will be resistent to the multiple timeout being scheduled.